ykman/piv.py

```diff
@@ -814,14 +815,27 @@ def sign_raw(self, slot, algorithm, message):
         return self._raw_sign_decrypt(slot, algorithm, Tlv(0x81, message),
                                       _sign_len_conditions[algorithm])
 
-    def sign(self, slot, algorithm, message):
+    def sign(self, slot, algorithm, message, pin=None, touch_callback=None):
+        message_hash = None
+
         if algorithm in (ALGO.RSA1024, ALGO.RSA2048):
-            message = _pkcs1_15_pad(algorithm, message)
+            message_hash = _pkcs1_15_pad(algorithm, message)
         elif algorithm in (ALGO.ECCP256, ALGO.ECCP384):
             h = hashes.Hash(hashes.SHA256(), default_backend())
             h.update(message)
-            message = h.finalize()
-        return self.sign_raw(slot, algorithm, message)
+            message_hash = h.finalize()
+        try:
+            return self.sign_raw(slot, algorithm, message_hash)
+        except APDUError as e:
+            if e.sw == SW.ACCESS_DENIED:
+                if pin:
+                    self.verify(pin, touch_callback=touch_callback)
+                    return self.sign(slot, algorithm, message, pin=pin)
```

Comment:
Isn't it wrong to pass the pin again to the recursive call? If it didn't work the first time...

Reply:
No, in fact it's required because - as you note above - the VERIFY command might follow up with an AUTHENTICATE command which consumes the verifiedness needed for the ALWAYS pin policy. If the PIN is wrong, then the `verify` call here will instead raise an exception that isn't caught here.

Reply:
So, if I'm understanding this correctly, the first call may get SW.ACCESS_DENIED due to lack of "verifiedness", which then calls verify(), and then a recursive call to sign, again passing pin.
However, the call to verify() might not provide the "verifiedness" the first time, as it might also do an authenticate(), which results in a second SW.ACCESS_DENIED. In that case a second recursive call to sign() is needed, making it a total of 3 invocations of sign() before success?

What about the `touch_callback`? That seems to get lost between the first and second calls to verify().

Reply:
Yes, that is a correct understanding.

>What about the `touch_callback`?

Oops, looks like I forgot about that one...